cvat/apps/engine/serializers.py

```diff
@@ -1364,51 +1439,201 @@ def update(self, instance: models.Task, validated_data: dict[str, Any]) -> model
+            # for some reason, but they can be seen in the profile
+            models.Image.objects.bulk_create(
+                list(bulk_context.updated_honeypots.values()),
+                update_conflicts=True,
+                update_fields=['path', 'width', 'height', 'real_frame'],
+                unique_fields=['id'],
+                batch_size=1000,
+            )
+
+            for updated_honeypots_batch in take_by(bulk_context.updated_honeypots, count=1000):
+                models.RelatedFile.images.through.objects.filter(
+                    image_id__in=updated_honeypots_batch
+                ).delete()
+
+            frame_provider = TaskFrameProvider(db_task)
+            validation_frame_uses_with_related_files: dict[int, list[int]] = {}
+            for honeypot_frame, db_honeypot in bulk_context.updated_honeypots.items():
+                validation_frame_uses_with_related_files.setdefault(
+                    frame_provider.get_rel_frame_number(db_honeypot.real_frame), []
+                ).append(honeypot_frame)
+
+            new_m2m_objects = []
+            related_files_m2m_objects = (
+                models.RelatedFile.images.through.objects
+                .filter(image_id__in=[
+                    bulk_context.task_frames[f].id for f in db_task.data.validation_layout.frames
+                ])
+                .all()
+            )
+            for validation_frame, validation_frame_uses in validation_frame_uses_with_related_files.items():
+                new_m2m_objects.extend(
+                    models.RelatedFile.images.through(
+                        image_id=bulk_context.task_frames[honeypot_frame].id,
+                        related_file_id=related_file_m2m_obj.related_file_id
+                    )
+                    for honeypot_frame in validation_frame_uses
+                    for related_file_m2m_obj in related_files_m2m_objects
+                    if related_file_m2m_obj.image_id == validation_frame
                 )
-                job_validation_layout_serializer.is_valid(raise_exception=True)
-                job_validation_layout_serializer.save()
 
-        return instance
+            models.RelatedFile.images.through.objects.bulk_create(new_m2m_objects, batch_size=1000)
+
+            updated_date = timezone.now()
+            for updated_segments_batch in take_by(updated_segments, count=1000):
+                models.Job.objects.filter(segment_id__in=updated_segments_batch).update(updated_date=updated_date)
+
+            for updated_segment_chunks_batch in take_by(bulk_context.segments_with_updated_chunks, count=1000):
+                models.Segment.objects.filter(id__in=updated_segment_chunks_batch).update(chunks_updated_date=updated_date)
+
+            db_task.touch()
+            if db_task.project:
+                db_task.project.touch()
+
+    def _clear_annotations_on_frames(self, db_task: models.Task, frames: Sequence[int]):
```

Comment:
Consider moving this function into models.py and managing which annotation types should be removed through arguments

Reply:
Done